byterun/shared_heap.c

```diff
@@ -143,25 +170,28 @@ void caml_sample_heap_stats(struct caml_heap_state* local, struct heap_stats* h)
 static pool* pool_acquire(struct caml_heap_state* local) {
   pool* r;
 
-  caml_plat_lock(&pool_freelist.lock);
-  if (!pool_freelist.free) {
-    void* mem = caml_mem_map(Bsize_wsize(POOL_WSIZE) * POOLS_PER_ALLOCATION,
+  do {
+    r = lockfree_queue_pop(&pool_freelist.free);
+
+    if (!r) {
+      void* mem = caml_mem_map(Bsize_wsize(POOL_WSIZE) * POOLS_PER_ALLOCATION,
                               Bsize_wsize(POOL_WSIZE), 0 /* allocate */);
-    int i;
-    if (mem) {
-      pool_freelist.free = mem;
-      for (i=1; i<POOLS_PER_ALLOCATION; i++) {
-        r = (pool*)(((uintnat)mem) + ((uintnat)i) * Bsize_wsize(POOL_WSIZE));
-        r->next = pool_freelist.free;
-        r->owner = 0;
-        pool_freelist.free = r;
+      int i;
+      if (mem) {
+        lockfree_queue_push(&pool_freelist.free, mem);
+        for (i=1; i<POOLS_PER_ALLOCATION; i++) {
+         r = (pool*)(((uintnat)mem) + ((uintnat)i) * Bsize_wsize(POOL_WSIZE));
+         r->owner = 0;
+         r->next = 0;
+         lockfree_queue_push(&pool_freelist.free, r);
+        }
+      } else {
+       return 0;
       }
     }
-  }
-  r = pool_freelist.free;
-  if (r)
-    pool_freelist.free = r->next;
-  caml_plat_unlock(&pool_freelist.lock);
+
+    r = lockfree_queue_pop(&pool_freelist.free);
```

Comment:
Why are we doing this twice (once on L174 and here on L193)? Will that not leak (or worse) if the first one returns a good pool (or is the last pool)?

Reply:
Another good find, this does leak. 193 is supposed to go between 190 and 191.